zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+///   it is not possible to derive new addresses for it.
+/// * `SqliteClientError::ReachedGapLimit`, if it is not possible to reserve `n` addresses
+///   within the gap limit after the last address in this account that is known to have an
+///   output in a mined transaction.
+/// * `SqliteClientError::AddressGeneration(AddressGenerationError::DiversifierSpaceExhausted)`,
+///   if the limit on transparent address indices has been reached.
+pub(crate) fn reserve_next_n_ephemeral_addresses<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    account_id: AccountId,
+    n: u32,
+) -> Result<Vec<(TransparentAddress, TransparentAddressMetadata)>, SqliteClientError> {
+    if n == 0 {
+        return Ok(vec![]);
+    }
+    assert!(n > 0);
+    let n = i32::try_from(n).expect("precondition violated");
+
+    let ephemeral_ivk = get_ephemeral_ivk(wdb.conn.0, &wdb.params, account_id)?;
+    let last_reserved_index = last_reserved_index(wdb.conn.0, account_id)?;
+    let last_safe_index = last_safe_index(wdb.conn.0, account_id)?;
+    let allocation = range_after(last_reserved_index, n);
+
+    if allocation.clone().count() < n.try_into().unwrap() {
+        return Err(SqliteClientError::AddressGeneration(
+            AddressGenerationError::DiversifierSpaceExhausted,
+        ));
+    }
+    if *allocation.end() > last_safe_index {
+        let unsafe_index = max(*allocation.start(), last_safe_index.saturating_add(1));
+        return Err(SqliteClientError::ReachedGapLimit(account_id, unsafe_index));
+    }
+
+    // used_in_tx and mined_in_tx are initially NULL
+    let mut stmt_insert_ephemeral_address = wdb.conn.0.prepare_cached(
+        "INSERT INTO ephemeral_addresses (account_id, address_index, address)
+         VALUES (:account_id, :address_index, :address)",
+    )?;
+
+    allocation
+        .map(|raw_index| {
+            let address_index = NonHardenedChildIndex::from_index(raw_index).unwrap();
+            let address = ephemeral_ivk.derive_ephemeral_address(address_index)?;
+
+            stmt_insert_ephemeral_address.execute(named_params![
+                ":account_id": account_id.0,
+                ":address_index": raw_index,
+                ":address": encode_transparent_address_p(&wdb.params, &address)
+            ])?;
+            Ok((address, metadata(address_index)))
+        })
+        .collect()
+}
+
+/// Returns a `SqliteClientError::EphemeralAddressReuse` error if `address` is
+/// an ephemeral transparent address.
+pub(crate) fn check_address_is_not_ephemeral<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    address_str: &str,
+) -> Result<(), SqliteClientError> {
+    ephemeral_address_check_internal(wdb, address_str, true)
+}
+
+/// Returns a `SqliteClientError::EphemeralAddressReuse` error if the address was
+/// already used. If `reject_all_ephemeral` is set, return an error if the address
+/// is ephemeral at all, regardless of reuse.
```

Comment:
This documentation is inaccurate: an error is only returned if the address is ephemeral _and_ has either been reserved, or detected as mined (i.e. the cases where `ephemeral_addresses` can contain a row). In particular:
- The gap limit is not checked to detect address ephemerality.
- We don't detect addresses that exist within internal gaps (that can occur in multi-wallet usage combined with expiring transactions), because AFAICT we don't fill those gaps in with rows when we detect a new address at a non-connecting index).
  - We don't currently _create_ internal gaps in this PR, but that is due to a bug (see later comment).

Update the documentation to reflect this, or alter the PR if the intention is for this to detect all ephemeral addresses up to the gap limit.

Reply:
This has been completely refactored, the `reject_all_ephemeral` flag no longer exists.